### PR TITLE
Handling delete Micro-gateway label when it is attached to an API

### PR DESCRIPTION
**Pull request by nalakas**
fixes for:
https://github.com/wso2/product-apim/issues/6069
https://github.com/wso2/product-apim/issues/6060
https://github.com/wso2/product-apim/issues/6057

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 1]
```diff
@@ -74,7 +76,10 @@ public Label addLabel(String tenantDomain, Label label) throws APIManagementExce
      * @param labelId Label identifier
      * @throws APIManagementException If failed to delete label
      */
-    public void deleteLabel(String labelId) throws APIManagementException{
+    public void deleteLabel(String user, String labelId) throws APIManagementException {
+        if (isAttachedLabel(user, labelId)) {
+            APIUtil.handleException("Unable to delete label. Label is already attached to a API");
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]:
```suggestion
            APIUtil.handleException("Unable to delete the label. It is attached to an API");
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 2]
```diff
@@ -106,6 +111,33 @@ public boolean isLableNameExists(String tenantDomain, Label label) throws APIMan
...
+        List<API> apiList = apiProvider.getAllAPIs();
+        List<Label> allLabelsWithID = getAllLabels(MultitenantUtils.getTenantDomain(user));
+        String labelName = null;
+        for (Label label : allLabelsWithID) {
+            if (labelId.equalsIgnoreCase(label.getLabelId())) {
+                labelName = label.getName();
+                break;
+            }
+        }
+        if (labelName != null && !labelName.isEmpty()) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: we can use strigutils.isempty()

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 3]
```diff
@@ -106,6 +111,33 @@ public boolean isLableNameExists(String tenantDomain, Label label) throws APIMan
...
+            UserAwareAPIProvider userAwareAPIProvider = new UserAwareAPIProvider(user);
+            for (API api : apiList) {
+                String uuid = api.getUUID();
+                API lightweightAPIByUUID = userAwareAPIProvider.getLightweightAPIByUUID(uuid, apiProvider.tenantDomain);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: more than 120 chars?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java [hunk 4]
```diff
@@ -53,11 +53,12 @@
...
         try {
             APIAdmin apiAdmin = new APIAdminImpl();
-            apiAdmin.deleteLabel(labelId);
+            String user = RestApiUtil.getLoggedInUsername();
+            apiAdmin.deleteLabel(user, labelId);
             return Response.ok().build();
         } catch (APIManagementException e) {
             String errorMessage = "Error while deleting API : " + labelId;
-            RestApiUtil.handleInternalServerError(errorMessage, e, log);
+            RestApiUtil.handleInternalServerError(errorMessage, log);
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: where do we log the error trace now?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 1 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 2 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 3 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java (hunk 4 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: we are logging 
APIUtil.handleException("Unable to delete label. Label is already attached to a API");

Do we need error trace anymore?

Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: where do we log that?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java [hunk 5]
```diff
@@ -53,11 +53,12 @@
     public Response labelsLabelIdDelete(String labelId, String ifMatch, String ifUnmodifiedSince) {
         try {
             APIAdmin apiAdmin = new APIAdminImpl();
-            apiAdmin.deleteLabel(labelId);
+            String user = RestApiUtil.getLoggedInUsername();
+            apiAdmin.deleteLabel(user, labelId);
             return Response.ok().build();
         } catch (APIManagementException e) {
             String errorMessage = "Error while deleting API : " + labelId;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: shouldn't this be "label"?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java [hunk 6]
```diff
@@ -104,7 +105,7 @@ public Response labelsPost(LabelDTO body) {
         } catch (APIManagementException | URISyntaxException e) {
             String errorMessage = "Error while adding new Label for API : "
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]:
```suggestion
            String errorMessage = "Error while adding new label to API : "
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java [hunk 7]
```diff
@@ -129,7 +130,7 @@ public Response labelsLabelIdPut(String labelId, LabelDTO body) {
             return Response.ok(location).entity(labelDTO).build();
         } catch (APIManagementException | URISyntaxException e) {
             String errorMessage = "Error while updating Label : " + labelId;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]:
```suggestion
            String errorMessage = "Error while updating label : " + labelId;
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java (hunk 4 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: in org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java (hunk 6 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java (hunk 5 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java (hunk 7 above)
Comment (nalakas) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/LabelsApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 1 above)
Comment (rmsamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: @nalakas Let's have the practice to add the commit hash too when mentioning an comment/issue/suggestion as fixed. i.e. Fixed with 46c6ec9

**Pull request merged**